modules/network/networker.go

```diff
@@ -1,27 +1,200 @@
+	"github.com/threefoldtech/zosv2/modules/network/bridge"
+	"github.com/threefoldtech/zosv2/modules/network/wireguard"
 
-// func (n *networker) GetNetResource(id string) (modules.NetResource, error) {
-// 	// TODO check signature
-// 	return n.netResAlloc.Get(id)
-// }
+	"github.com/threefoldtech/zosv2/modules/network/namespace"
 
-// func (n *networker) ApplyNetResource(resource modules.NetResource) error {
-// 	return nil
-// }
+	"github.com/threefoldtech/zosv2/modules"
+)
 
-// type NetResourceAllocator interface {
-// 	Get(txID string) (modules.NetResource, error)
-// }
+type networker struct {
+	storageDir  string
+	netResAlloc NetResourceAllocator
+}
+
+func NewNetworker(storageDir string, allocator NetResourceAllocator) modules.Networker {
+	return &networker{
+		storageDir:  storageDir,
+		netResAlloc: allocator,
+	}
+}
+
+var _ modules.Networker = (*networker)(nil)
+
+func (n *networker) GetNetResource(id string) (modules.NetResource, error) {
+	// TODO check signature
+	return n.netResAlloc.Get(id)
+}
+
+func (n *networker) ApplyNetResource(netID modules.NetID, resource modules.NetResource) error {
+	return applyNetResource(n.storageDir, netID, resource)
+}
+
+func applyNetResource(storageDir string, netID modules.NetID, netRes modules.NetResource) error {
+	network := string(netID)
+	if err := createNetwork(network); err != nil {
+		return err
+	}
+
+	storage := filepath.Join(storageDir, network)
+	if err := configureWG(storage, network, netRes); err != nil {
+		return err
+	}
+	return nil
+}
+
+// createNetwork creates a network namespace and a bridge
+// and a wireguard interface and then move it interface inside
+// the net namespace
+func createNetwork(name string) error {
+	var (
+		netnsName  = netnsName(name)
+		bridgeName = bridgeName(name)
+		wgName     = wgName(name)
+	)
+
+	if !namespace.Exists(netnsName) {
+		log.Info().Str("name", netnsName).Msg("create network namespace")
+		_, err := namespace.Create(netnsName)
+		if err != nil {
+			return err
+		}
+
+		log.Info().Str("name", wgName).Msg("create wireguard interface")
+		wg, err := wireguard.New(wgName)
+		if err != nil {
+			return err
+		}
+
+		log.Info().
+			Str("wg", wgName).
+			Str("namespace", netnsName).
+			Msg("move wireguard into network namespace")
+		if err := namespace.SetLink(wg, netnsName); err != nil {
+			return err
+		}
+	}
+
+	if !bridge.Exists(bridgeName) {
+		if _, err := bridge.New(bridgeName); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func deleteNetwork(network string) error {
+	if err := bridge.Delete(bridgeName(network)); err != nil {
+		return err
+	}
+	return namespace.Delete(netnsName(network))
+}
+
+func configureWG(storageDir, network string, netRes modules.NetResource) error {
+	var (
+		netnsName   = netnsName(network)
+		wgName      = wgName(network)
+		storagePath = filepath.Join(storageDir, network)
+		key         wgtypes.Key
+		err         error
+	)
+
+	key, err = wireguard.LoadKey(storagePath)
```

Comment:
what if the error is a storage error ?

Reply:
Well in any case, If I cannot read the key from the disk, would it be because it has not been generated before or storage error, I need to create a new one.
But this means I need a way to update the resource unit when things like that happens, cause all the other node will have to upgrade their wg config with the new public key.